bpf/bpf_lxc.c

```diff
@@ -351,11 +360,15 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 		key.ip6.p4 = 0;
 		key.family = ENDPOINT_KEY_IPV6;
 
-		ret = encap_and_redirect(skb, &key, SECLABEL, monitor, true);
-
-		/* Fall through if remote prefix was not found
-		 * (DROP_NO_TUNNEL_ENDPOINT) */
-		if (ret != DROP_NO_TUNNEL_ENDPOINT)
+		/* Three cases exist here either (a) the encap and redirect could
+		 * not find the tunnel so fallthrough to nat46 and stack, (b)
+		 * the packet needs IPSec encap so push skb to stack for encap, or
+		 * (c) packet was redirected to tunnel device so return.
+		 */
+		ret = encap_and_redirect(skb, &key, SECLABEL, monitor, false);
+		if (ret == IPSEC_ENDPOINT)
+			goto pass_to_stack;
+		else if (ret != DROP_NO_TUNNEL_ENDPOINT)
```

Comment:
Nit: This is not 100% consistent with the IPv4 path. I think they're both functionally very similar, but they are written slightly differently (ordering, number of paths, fallthrough to the `goto` immediately following the current block).

IPv4 path for comparison:
```
                ret = encap_and_redirect(skb, &key, SECLABEL, monitor, false);  
                if (ret == DROP_NO_TUNNEL_ENDPOINT)                             
                        goto pass_to_stack;                                     
                else if (ret == IPSEC_ENDPOINT)                                 
                        goto pass_to_stack;                                     
                else                                                            
                        return ret;      
```

Reply:
Agree there is a bit of lack of foresight here from where I added the V4 version. The IPv6 variant needs to be like above to fallthrough to the NAT code path that is before the `goto pass_to_stack`. The IPv4 variant could take the form of the IPv6 version but not the other way around. To unify them we would need another patch on the v4 side to make them similar. I could try to argue that somehow keeping the goto if/else is better for code gen but I can't actually see how that would be true :/

Reply:
I see. If we can avoid this with an additional patch on top, without otherwise breaking eg verifier on v4.9 then it'd be nice to make the paths more consistent. This can be addressed in a follow-up PR if that makes it any easier.

Reply:
Yeah lets push into a follow up if no one minds. Its a bit orthogonal to enabling ipv6. Nice catch though.